test/framework/machine_helpers.go

```diff
@@ -21,32 +21,106 @@ import (
+// it is necessary to ensure this is already happened before calling it.
+func GetControlPlaneMachinesByCluster(ctx context.Context, input GetControlPlaneMachinesByClusterInput) []clusterv1.Machine {
+	options := append(byClusterOptions(input.ClusterName, input.Namespace), controlPlaneMachineOptions()...)
+
+	machineList := &clusterv1.MachineList{}
+	Expect(input.Lister.List(ctx, machineList, options...)).To(Succeed(), "Failed to list MachineList object for Cluster %s/%s", input.Namespace, input.ClusterName)
+
+	return machineList.Items
+}
+
+type GetMDMachinesByClusterInput struct {
+	GetMachinesByClusterInput
+	Deployment clusterv1.MachineDeployment
+}
+
+// GetMDMachinesByCluster returns Machine objects for a cluster belonging to a machine deployment.
+// Important! this method relies on labels that are created by the CAPI controllers during the first reconciliation, so
+// it is necessary to ensure this is already happened before calling it.
+func GetMDMachinesByCluster(ctx context.Context, input GetMDMachinesByClusterInput) []clusterv1.Machine {
+	opts := byClusterOptions(input.ClusterName, input.Namespace)
+	opts = append(opts, machineDeploymentOptions(input.Deployment)...)
+
+	machineList := &clusterv1.MachineList{}
+	Expect(input.Lister.List(ctx, machineList, opts...)).To(Succeed(), "Failed to list MachineList object for Cluster %s/%s", input.Namespace, input.ClusterName)
+
+	return machineList.Items
+}
+
 // WaitForMachinesToBeUpgradedInput is the input for WaitForMachinesToBeUpgraded.
 type WaitForMachinesToBeUpgradedInput struct {
-	Lister                   Lister
-	Cluster                  *clusterv1.Cluster
-	KubernetesUpgradeVersion string
-	MachineCount             int
+	Lister         Lister
+	Cluster        *clusterv1.Cluster
+	UpgradeVersion string
+	MachineCount   int
+}
+
+// WaitForControlPlaneMachinesToBeUpgraded waits until all control plane machines are upgraded to the correct kubernetes version.
+func WaitForControlPlaneMachinesToBeUpgraded(ctx context.Context, input WaitForMachinesToBeUpgradedInput, intervals ...interface{}) {
+	By("ensuring all machines have upgraded kubernetes version")
+	Eventually(func() (int, error) {
+		machines := GetControlPlaneMachinesByCluster(context.TODO(), GetControlPlaneMachinesByClusterInput{
+			GetMachinesByClusterInput: GetMachinesByClusterInput{
+				Lister:      input.Lister,
+				ClusterName: input.Cluster.Name,
+				Namespace:   input.Cluster.Namespace,
+			},
+		})
+
+		upgraded := 0
+		for _, machine := range machines {
+			if *machine.Spec.Version == input.UpgradeVersion {
+				upgraded++
+			}
+		}
+		if len(machines) > upgraded {
+			return 0, errors.New("old nodes remain")
+		}
+		return upgraded, nil
+	}, intervals...).Should(Equal(input.MachineCount))
+}
+
+// WaitForMDMachinesToBeUpgradedInput is the input for WaitForMDMachinesToBeUpgraded.
+type WaitForMDMachinesToBeUpgradedInput struct {
```

Comment:
this should be WaitForMachineDeploymentsMachinesToBeUpgradedInput (this is verbose, but this is the pattern we are following)